### PR TITLE
*: support stream when read cluster memory table.

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
Before this PR, TiDB doesn't support reading cluster memory table with `stream`.

### What is changed and how it works?
* Implement `CoprocessorStream` interface for `rpc_server` to support stream.
* Turn on stream when reading cluster memory table.

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

Code changes

 - Has exported function/method change

Side effects

### CODE REVIEW

File: executor/coprocessor.go [hunk 1]
```diff
@@ -70,12 +68,66 @@ func (h *CoprocessorDAGHandler) HandleRequest(ctx context.Context, req *coproces
...
+	err = e.Open(ctx)
+	if err != nil {
+		return stream.Send(h.buildErrorResponse(err))
+	}
+
+	chk := newFirstChunk(e)
+	tps := e.base().retFieldTypes
+	for {
+		chk.Reset()
+		err = Next(ctx, e, chk)
```
Comment (bb7133) [inline executor/coprocessor.go]:
```suggestion
		if err = Next(ctx, e, chk); err != nil {...}
```

File: executor/coprocessor.go [hunk 2]
```diff
@@ -70,12 +68,66 @@ func (h *CoprocessorDAGHandler) HandleRequest(ctx context.Context, req *coproces
...
+
+func (h *CoprocessorDAGHandler) buildResponseAndSendToStream(chk *chunk.Chunk, tps []*types.FieldType, stream tikvpb.Tikv_CoprocessorStreamServer) error {
+	chunks, err := h.buildChunk(chk, tps)
+	if err != nil {
+		return stream.Send(h.buildErrorResponse(err))
+	}
+
+	for _, c := range chunks {
+		resp := h.buildStreamResponse(&c)
+		err = stream.Send(resp)
```
Comment (bb7133) [inline executor/coprocessor.go]: Ditto

File: server/rpc_server.go [hunk 3]
```diff
@@ -89,6 +92,31 @@ func (s *rpcServer) Coprocessor(ctx context.Context, in *coprocessor.Request) (r
 	return resp, nil
 }
 
+// Coprocessor implements the TiKVServer interface.
+func (s *rpcServer) CoprocessorStream(in *coprocessor.Request, stream tikvpb.Tikv_CoprocessorStreamServer) (err error) {
+	resp := &coprocessor.Response{}
+	defer func() {
+		if v := recover(); v != nil {
+			logutil.BgLogger().Error("panic in TiDB RPC server coprocessor", zap.Any("stack", v))
+			resp.OtherError = fmt.Sprintf("rpc coprocessor panic, :%v", v)
```
Comment (bb7133) [inline server/rpc_server.go]: Please improve this error message.

File: store/mockstore/mocktikv/rpc.go [hunk 4]
```diff
@@ -723,9 +733,21 @@ func (c *RPCClient) checkArgs(ctx context.Context, addr string) (*rpcHandler, er
 	return handler, nil
 }
 
-// TiDBRPCServerCoprocessorHandler is the TiDB rpc server coprocessor handler.
+// GRPCClientFactor is the GRPC client factory.
+// Use global variable to avoid circle import.
 // TODO: remove this global variable.
-var TiDBRPCServerCoprocessorHandler func(context.Context, *coprocessor.Request) *coprocessor.Response
+var GRPCClientFactor func() Client
```
Comment (bb7133) [inline store/mockstore/mocktikv/rpc.go]: Why naming it 'Factor'?

File: server/rpc_server.go [hunk 5]
```diff
@@ -89,6 +92,31 @@ func (s *rpcServer) Coprocessor(ctx context.Context, in *coprocessor.Request) (r
 	return resp, nil
 }
 
+// Coprocessor implements the TiKVServer interface.
```
Comment (bb7133) [inline server/rpc_server.go]: Please refine this comment

File: store/mockstore/mocktikv/rpc.go [hunk 6]
```diff
@@ -667,13 +667,23 @@ func (h *rpcHandler) handleSplitRegion(req *kvrpcpb.SplitRegionRequest) *kvrpcpb
...
+}
+
 // RPCClient sends kv RPC calls to mock cluster. RPCClient mocks the behavior of
 // a rpc client at tikv's side.
 type RPCClient struct {
 	Cluster       *Cluster
 	MvccStore     MVCCStore
 	streamTimeout chan *tikvrpc.Lease
 	done          chan struct{}
+	rpcCli        Client
```
Comment (bb7133) [inline store/mockstore/mocktikv/rpc.go]: This is weird, any way to avoid this?

File: executor/coprocessor.go [hunk 7]
```diff
@@ -70,12 +68,66 @@ func (h *CoprocessorDAGHandler) HandleRequest(ctx context.Context, req *coproces
 		if chk.NumRows() == 0 {
 			break
 		}
-		err = h.appendChunk(chk, tps)
+		partChunks, err = h.buildChunk(chk, tps)
 		if err != nil {
 			break
```
Comment (wjhuang2016) [inline executor/coprocessor.go]: How about `return h.buildErrorResponse(err)` instead of `break`?

File: store/mockstore/mocktikv/rpc.go (hunk 4 above)
Comment (crazycs520) [reply store/mockstore/mocktikv/rpc.go]: Sorry, It's a mistake, It should be `factory`.

File: store/mockstore/mocktikv/rpc.go (hunk 6 above)
Comment (crazycs520) [reply store/mockstore/mocktikv/rpc.go]: em... It was the simplest way to redirect the request and avoid circle import problems that I have thought...

File: executor/coprocessor.go (hunk 7 above)
Comment (crazycs520) [reply executor/coprocessor.go]: great, done.

File: server/rpc_server.go (hunk 3 above)
Comment (crazycs520) [reply server/rpc_server.go]: done.

File: store/mockstore/mocktikv/rpc.go [hunk 8]
```diff
@@ -723,9 +733,21 @@ func (c *RPCClient) checkArgs(ctx context.Context, addr string) (*rpcHandler, er
...
 
-// TiDBRPCServerCoprocessorHandler is the TiDB rpc server coprocessor handler.
+// GRPCClientFactory is the GRPC client factory.
+// Use global variable to avoid circle import.
 // TODO: remove this global variable.
-var TiDBRPCServerCoprocessorHandler func(context.Context, *coprocessor.Request) *coprocessor.Response
+var GRPCClientFactory func() Client
+
+func (c *RPCClient) redirectRequestToRPCServer(ctx context.Context, addr string, req *tikvrpc.Request, timeout time.Duration) (*tikvrpc.Response, error) {
+	if c.rpcCli == nil {
```
Comment (bb7133) [inline store/mockstore/mocktikv/rpc.go]: I guess this function will be called concurrently?

File: executor/coprocessor.go [hunk 9]
```diff
@@ -32,50 +33,93 @@ import (
...
+		if chk.NumRows() == 0 {
 			break
 		}
+		err = h.buildResponseAndSendToStream(chk, tps, stream)
+		if err != nil {
+			return stream.Send(h.buildErrorResponse(err))
+		}
 	}
-	return h.buildResponse(err)
+	return h.buildResponseAndSendToStream(chk, tps, stream)
```
Comment (wjhuang2016) [inline executor/coprocessor.go]: This line can move into line 100, and use return.

File: infoschema/tables_test.go [hunk 10]
```diff
@@ -1031,29 +1035,32 @@ func (s *testClusterTableSuite) TestSelectClusterTable(c *C) {
...
-	c.Assert(re, NotNil)
-	c.Assert(len(re.Rows()) > 0, IsTrue)
-	tk.MustExec("set @@global.tidb_enable_stmt_summary=0")
-	re = tk.MustQuery("select * from `CLUSTER_events_statements_summary_by_digest`")
-	c.Assert(re, NotNil)
-	c.Assert(len(re.Rows()) == 0, IsTrue)
-	tk.MustQuery("select * from `CLUSTER_events_statements_summary_by_digest_history`")
-	c.Assert(re, NotNil)
-	c.Assert(len(re.Rows()) == 0, IsTrue)
+	for i := 0; i < 2; i++ {
```
Comment (wjhuang2016) [inline infoschema/tables_test.go]: Why do this?

File: server/rpc_server.go [hunk 11]
```diff
@@ -81,14 +84,39 @@ func (s *rpcServer) Coprocessor(ctx context.Context, in *coprocessor.Request) (r
 	resp = &coprocessor.Response{}
 	defer func() {
 		if v := recover(); v != nil {
-			logutil.BgLogger().Error("panic in TiDB RPC server coprocessor", zap.Any("stack", v))
-			resp.OtherError = fmt.Sprintf("rpc coprocessor panic, :%v", v)
+			logutil.BgLogger().Error("panic when when RPC server handing coprocessor", zap.Any("stack", v))
```
Comment (wjhuang2016) [inline server/rpc_server.go]: `when when`?

File: store/mockstore/mocktikv/rpc.go (hunk 6 above)
Comment (bb7133) [reply store/mockstore/mocktikv/rpc.go]: Then please add some detailed comments here to explain why we decide to put a `rpcCli` here.

File: infoschema/tables_test.go (hunk 10 above)
Comment (crazycs520) [reply infoschema/tables_test.go]: https://github.com/pingcap/tidb/pull/14344/files#diff-9cb2947f24dfdff2e6f6cec84b53aa56R1040
Test for no stream and with stream.

File: store/mockstore/mocktikv/rpc.go (hunk 8 above)
Comment (crazycs520) [reply store/mockstore/mocktikv/rpc.go]: Great, done.

File: executor/coprocessor.go (hunk 9 above)
Comment (crazycs520) [reply executor/coprocessor.go]: done.

File: server/rpc_server.go (hunk 11 above)
Comment (crazycs520) [reply server/rpc_server.go]: great, done.

File: executor/coprocessor.go [hunk 12]
```diff
@@ -107,68 +150,84 @@ func (h *CoprocessorDAGHandler) buildDAGExecutor(req *coprocessor.Request) (Exec
...
 	default:
-		return errors.Errorf("unknown DAG encode type: %v", h.dagReq.EncodeType)
+		return nil, errors.Errorf("unknown DAG encode type: %v", h.dagReq.EncodeType)
 	}
-	return err
+	return chunks, err
 }
 
-func (h *CoprocessorDAGHandler) buildResponse(err error) *coprocessor.Response {
+func (h *CoprocessorDAGHandler) buildResponse(chunks []tipb.Chunk) *coprocessor.Response {
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply executor/coprocessor.go]: Why name `unary` response? Do we have a `multivariate ` response?

**Pull request merged**